### PR TITLE
GSoC: Add Proposal for KubeBuilder - Phase 2 Plugins API 

### PR DESCRIPTION
**Pull request by rashmigottipati**
Would like to add this proposal to summerofcode 2023.

### CODE REVIEW

File: summerofcode/2023.md [hunk 1]
```diff
@@ -382,7 +382,30 @@ OCI Artifacts are a set of conventions that allows us to store assets other than
 
 ### Kubebuilder
 
+#### Leverage Phase 2 Plugins API to allow creation and usage of external plugins
```
Comment (nate-double-u) [inline summerofcode/2023.md:385]: Could you add this to the table of contents above?

**Review by nate-double-u — COMMENTED** [review]
Thanks for the proposal @rashmigottipati.

I'm looking at the `owners_aliases` file for kubebuilder, and see both the mentors under the `kubebuilder-reviewers` category. Does that put you in the maintainer role for the sub-project, or does someone like @camilamacedo86 in the `kubebuilder-admins` list need to provide a thumbs up? (Sorry i'm not very familiar with the KubeBuilder maintainer structure)

File: summerofcode/2023.md (hunk 1 above)
Comment (nate-double-u) [reply summerofcode/2023.md:385]: Thanks for the  clarification! We can get this in still, just need the Table of Contents update, and then we can get these two into the system

File: summerofcode/2023.md [hunk 2]
```diff
@@ -382,7 +382,30 @@ OCI Artifacts are a set of conventions that allows us to store assets other than
...
+
+- Description: Kubebuilder introduced Phase 2 Plugins to make itself more extensible as a CLI and as a library. This enables the community to develop, use, maintain and release their own plugins independently. Phase 2 Plugins has been available for a few releases, and in order to further increase and help facilitate adoption of this functionality, we need to work on creating an extensive sample, documentation and tutorial on how to build and use external plugins. To be able to achieve this goal, we can use go, ansible, python or helm to create the sample external plugin that will be used as a central reference plugin that serves as an example of an external plugin that leverages Phase 2 Plugins.
+
+- Expected outcome:
+  - Finish the meta task: <https://github.com/kubernetes-sigs/kubebuilder/issues/2600>
```
Comment (aliok) [inline summerofcode/2023.md:390]:
```suggestion
  - Finish the remaining tasks in the meta task (the big task that tracks various subtasks for the feature): <https://github.com/kubernetes-sigs/kubebuilder/issues/2600>
```

You might want to explain what a meta task is, like above, if you like, considering that potential mentees might not know what it means.

File: summerofcode/2023.md [hunk 3]
```diff
@@ -382,7 +382,30 @@ OCI Artifacts are a set of conventions that allows us to store assets other than
...
+      - Adding e2e tests which should validate that the sample can be built and used with the Kubebuilder CLI.
+      - Implementing a maintainable solution to ensure that the sample will not get outdated.
+        - We might need to so something like: <https://github.com/kubernetes-sigs/kubebuilder/pull/3191>
+        - However, a dependabot for the sample might suffice in this case, just to ensure the dependencies.
+    - Review and improve the Plugin documentation to ensure that we the community can easily leverage the Phase 2 Plugins API: <https://book.kubebuilder.io/plugins/plugins.html>
+    - Verify that we are generating go docs for the Phase 2 Plugins API.
+    - Create a tutorial with the sample to let the community and users know how to build and use their own external plugins.
+    - Add an extra sample using any other language other than Golang.
+    - Create a video demo to showcase the Phase 2 Plugin working with the sample. This will demonstrate the achievements and will be shared with the community. We also will add the link to the demo in the docs.
+- Recommended Skills: Go
```
Comment (aliok) [inline summerofcode/2023.md:401]:
```suggestion
- Recommended Skills: Go and one of Ansible, Python or Helm
```

I saw references to  go, ansible, python or helm above. You might want to list them here, but it is ok to leave them out as well if knowledge of these are not critical.

**Review by aliok — COMMENTED** [review]
Thanks for the proposal @rashmigottipati 

Added 2 minor suggestions. Feel free to incorporate them in.

As Nate mentioned, table of contents needs an update.

File: summerofcode/2023.md (hunk 3 above)
Comment (camilamacedo86) [reply summerofcode/2023.md:401]: I think only Go is required @rashmigottipati @everettraven WDYT?

File: summerofcode/2023.md [hunk 4]
```diff
@@ -382,7 +382,30 @@ OCI Artifacts are a set of conventions that allows us to store assets other than
 
 ### Kubebuilder
 
+#### Leverage Phase 2 Plugins API to allow creation and usage of external plugins
+
+- Description: Kubebuilder introduced Phase 2 Plugins to make itself more extensible as a CLI and as a library. This enables the community to develop, use, maintain and release their own plugins independently. Phase 2 Plugins has been available for a few releases, and in order to further increase and help facilitate adoption of this functionality, we need to work on creating an extensive sample, documentation and tutorial on how to build and use external plugins. To be able to achieve this goal, we can use go, ansible, python or helm to create the sample external plugin that will be used as a central reference plugin that serves as an example of an external plugin that leverages Phase 2 Plugins.
```
Comment (camilamacedo86) [inline summerofcode/2023.md:387]:
```suggestion
- Description: KubeBuilder introduced [Phase 2 Plugins](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/extensible-cli-and-scaffolding-plugins-phase-2.md) to make itself more extensible as a CLI and as a library. This enables the community to develop, use, maintain and release their own plugins independently. Phase 2 Plugins has been available for a few releases, and in order to further increase and help facilitate adoption of this functionality, we need to work on creating an extensive sample, documentation and tutorial on how to build and use external plugins. To be able to achieve this goal, we can use Go, Ansible, Python or Helm to create the sample external plugin that will be used as a central reference plugin that serves as an example of an external plugin that leverages Phase 2 Plugins.
```

File: summerofcode/2023.md (hunk 3 above)
Comment (rashmigottipati) [reply summerofcode/2023.md:401]: As a required skill, I think Go is good enough. Other skills are optional in my opinion.

File: summerofcode/2023.md (hunk 1 above)
Comment (rashmigottipati) [reply summerofcode/2023.md:386]: Added to the table of contents. Addressed in latest commit. Thanks!

**Comment by rashmigottipati** [comment]
@nate-double-u @aliok It's ready for review again. Thanks for the review feedback!

/cc @camilamacedo86 @everettraven

File: summerofcode/2023.md [hunk 5]
```diff
@@ -61,6 +61,7 @@ If you are a project maintainer and consider mentoring during the GSoC 2023 cycl
     + [GitHub Actions for KubeArmor](#github-actions-for-kubearmor)
     + [Store KubeArmor policies in OCI registry](#store-kubearmor-policies-in-oci-registry)
   * [Kubebuilder](#kubebuilder)
+    + [Leverage Phase 2 Plugins API to allow creation and usage of external plugins](#leverage-phase2-plugins-api)
```
Comment (aliok) [inline summerofcode/2023.md:64]:
```suggestion
    + [Leverage Phase 2 Plugins API to allow creation and usage of external plugins](##leverage-phase-2-plugins-api-to-allow-creation-and-usage-of-external-plugins)
```

Comment (aliok) [reply summerofcode/2023.md:64]: Let's get fix the anchor and we're good to go

Comment (rashmigottipati) [reply summerofcode/2023.md]: Fixed it.

**Review by aliok — APPROVED** [review]
Thanks!
/lgtm
/approve

**Comment by aliok** [comment]
@rashmigottipati 
I saw you're already in CNCF GSoC organization. Please visit https://summerofcode.withgoogle.com/. Google asks you to accept the program terms. Then we can add you as a mentor for 2023 program.

@everettraven
I just sent you an invite for CNCF GSoC program. Please accept it :)

**Pull request merged**